File: ol-objc/ol-objc/OLValue.m
//
//  OLValue.m
//  ol-objc
//
//  Created by 伍 威 on 15/1/8.
//  Copyright (c) 2015年 sleepsaint. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "OLValue.h"
#import "OLSourceValue.h"

@implementation NSString (OLValue)

- (NSComparisonResult) compare2:(id)other {
    if ([other isKindOfClass:NSString.class]) {
        return [self compare:other];
    } else if ([other isKindOfClass:NSNumber.class]) {
        return [[NSNumber numberWithDouble:self.doubleValue] compare:other];
    } else {
        return 0;
    }
}
@end

@implementation NSNumber (OLValue)

- (NSComparisonResult) compare2:(id)other {
    if ([other isKindOfClass:NSNumber.class]) {
        return [self compare:other];
    } else if ([other isKindOfClass:NSString.class]) {
        return [self compare:[NSNumber numberWithDouble:[other doubleValue]]];
    } else {
        return 0;
    }
}
@end

@implementation NSArray (OLValue)

- (id) valueForKey:(NSString *)key {
    return [self objectAtIndex:key.integerValue];
}

- (void) setValue:(id)value forKey:(NSString *)key {
    NSMutableArray* array = (NSMutableArray*)self;
    [array replaceObjectAtIndex:key.integerValue withObject:value];
}

- (void)removeObjectForKey:(NSString *)key {
    NSMutableArray* array = (NSMutableArray*)self;
    [array removeObjectAtIndex:key.integerValue];
}

- (id) filter:(id)function root:(id)root temp:(id)temp {
    NSMutableArray* ret = [NSMutableArray arrayWithCapacity:self.count];
    for (id now in self) {
        if ([function lookup:root temp:temp now:now]) {
            [ret addObject:now];
        }
    }
    return ret;
}
- (id) some:(id)function root:(id)root temp:(id)temp {
    for (id now in self) {
        if ([[function lookup:root temp:temp now:now] boolValue]) {
            return [NSNumber numberWithBool:YES];
        }
    }
    return [NSNumber numberWithBool:NO];
}
- (NSArray*) arrayValue {
    return self;
}
@end

@implementation NSDictionary (OLValue)

- (id) filter:(id)function root:(id)root temp:(id)temp {
    NSMutableDictionary* ret = [NSMutableDictionary dictionaryWithCapacity:self.count];
    for (id key in self) {
        if ([[function lookup:root temp:temp now:[self objectForKey:key]] boolValue]) {
            ret[key] = [self objectForKey:key];
        }
    }
    return ret;
}
- (id) some:(id)function root:(id)root temp:(id)temp {
    for (id key in self) {
        if ([[function lookup:root temp:temp now:[self objectForKey:key]] boolValue]) {
            return [NSNumber numberWithBool:YES];
        }
    }
    return [NSNumber numberWithBool:NO];
}

- (NSArray*) arrayValue {
    return self.allValues;
}
@end

@implementation NSNull (OLValue)

- (NSString*) stringValue {
    return @"null";
}
- (id) filter:(id)function root:(id)root temp:(id)temp {
    return nil;
}
@end

File: ol-swift/ol-swift/OLValue.swift
//
//  OLValue.swift
//  ol-swift
//
//  Created by 伍 威 on 15/1/7.
//  Copyright (c) 2015年 sleepsaint. All rights reserved.
//

import Foundation

protocol OLValue {
    func toBool() -> Bool
    func getValueByKey(key: OLValue) -> OLValue
    func compare(b: OLValue) -> Int
    func filter(function: OLList, root: OLValue, temp: OLValue) -> OLValue
    func some(function: OLList, root: OLValue, temp: OLValue) -> Bool
    func toArray() -> [AnyObject]
}

func OLCompare(a: Double, b: Double) -> Int {
    if a > b {
        return 1
    } else if a < b {
        return -1
    } else {
        return 0
    }
}

func OLCompare(a: Int, b: Int) -> Int {
    if a > b {
        return 1
    } else if a < b {
        return -1
    } else {
        return 0
    }
}

func OLCompare(a: String, b: String) -> Int {
    if a > b {
        return 1
    } else if a < b {
        return -1
    } else {
        return 0
    }
}

extension NSNumber: OLValue {
    func toBool() -> Bool {
        return doubleValue != 0
    }
    func getValueByKey(key: OLValue) -> OLValue {
        return NSNull()
    }
    func compare(b: OLValue) -> Int {
        switch b {
        case let n as NSNumber:
            return OLCompare(doubleValue, n.doubleValue)
        case let s as NSString:
            return OLCompare(doubleValue, s.doubleValue)
        default:
            return 0
        }
    }
    func filter(function: OLList, root: OLValue, temp: OLValue) -> OLValue {
        return self
    }
    func some(function: OLList, root: OLValue, temp: OLValue) -> Bool {
        return true
    }
    func toArray() -> [AnyObject] {
        return []
    }
}

extension NSString: OLValue {
    func toBool() -> Bool {
        return length > 0
    }
    func getValueByKey(key: OLValue) -> OLValue {
        return NSNull()
    }
    func compare(b: OLValue) -> Int {
        switch b {
        case let n as NSNumber:
            return OLCompare(doubleValue, n.doubleValue)
        case let s as NSString:
            return OLCompare(self, s)
        default:
            return 0
        }
    }
    func filter(function: OLList, root: OLValue, temp: OLValue) -> OLValue {
        return self
    }
    func some(function: OLList, root: OLValue, temp: OLValue) -> Bool {
        return true
    }
    func toArray() -> [AnyObject] {
        return []
    }
}

extension NSArray: OLValue {
    func toBool() -> Bool {
        return count > 0
    }
    func getValueByKey(key: OLValue) -> OLValue {
        switch key {
        case let a as NSString:
            return toOLValue(objectAtIndex(a.integerValue))
        case let a as NSNumber:
            return toOLValue(objectAtIndex(a.integerValue))
        default:
            return NSNull()
        }
    }
    func compare(b: OLValue) -> Int {
        switch b {
        case let n as NSArray:
            return OLCompare(count, n.count)
        default:
            return 0
        }
    }
    func filter(function: OLList, root: OLValue, temp: OLValue) -> OLValue {
        let ret = NSMutableArray();
        for i in self {
            if function.lookup(root, temp: temp, now: toOLValue(i)).toBool() {
                ret.addObject(i)
            }
        }
        return ret
    }
    func some(function: OLList, root: OLValue, temp: OLValue) -> Bool {
        for i in self {
            if function.lookup(root, temp: temp, now: toOLValue(i)).toBool() {
                return true
            }
        }
        return false
    }
    func toArray() -> [AnyObject] {
        return self
    }
}

extension NSDictionary: OLValue {
    func toBool() -> Bool {
        return count > 0
    }
    func getValueByKey(key: OLValue) -> OLValue {
        switch key {
        case let a as NSString:
            return toOLValue(objectForKey(a))
        default:
            return NSNull()
        }
    }
    func compare(b: OLValue) -> Int {
        switch b {
        case let n as NSDictionary:
            return OLCompare(count, n.count)
        default:
            return 0
        }
    }
    func filter(function: OLList, root: OLValue, temp: OLValue) -> OLValue {
        let ret = NSMutableDictionary();
        for (key, i) in self {
            if function.lookup(root, temp: temp, now: toOLValue(i)).toBool() {
                ret[key as NSString] = i
            }
        }
        return ret
    }
    func some(function: OLList, root: OLValue, temp: OLValue) -> Bool {
        for (key, i) in self {
            if function.lookup(root, temp: temp, now: toOLValue(i)).toBool() {
                return true
            }
        }
        return false
    }
    func toArray() -> [AnyObject] {
        return allValues
    }
}

extension NSNull: OLValue {
    func toBool() -> Bool {
        return false
    }
    func getValueByKey(key: OLValue) -> OLValue {
        return NSNull()
    }
    func compare(b: OLValue) -> Int {
        switch b {
        case let n as NSNull:
            return 0
        default:
            return -1
        }
    }
    func filter(function: OLList, root: OLValue, temp: OLValue) -> OLValue {
        return self
    }
    func some(function: OLList, root: OLValue, temp: OLValue) -> Bool {
        return false
    }
    func toArray() -> [AnyObject] {
        return []
    }
}

func toOLValue(a: AnyObject?) -> OLValue {
    switch a {
    case let b as NSNumber:
        return b
    case let b as NSString:
        return b
    case let b as NSArray:
        return b
    case let b as NSDictionary:
        return b
    default:
        return NSNull()
    }
}


